gci/src/main/java/com/gcinterceptor/gci/GarbageCollectorControlInterceptor.java

```diff
@@ -0,0 +1,115 @@
+package com.gcinterceptor.gci;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class GarbageCollectorControlInterceptor {
+	private static final Duration WAIT_FOR_TRAILERS_SLEEP_MILLIS = Duration.ofMillis(10);
+	private static final int SAMPLE_HISTORY_SIZE = 5;
+	private AtomicBoolean doingGC;
+	private AtomicLong incoming;
+	private AtomicLong finished;
+	private AtomicLong shedRequests;
+	private Heap monitor;
+	private SheddingThreshold sheddingThreshold;
+	private Sampler sampler;
+	private Executor executor;
+	
+	/**
+	 * Creates a new instance of {@code GarbageCollectorControlInterceptor}
+	 *
+	 * @param monitor
+	 *            {@code HeapMonitor} used to monitoring JVM heap pools.
+	 * @param executor
+	 *            thread pool used to trigger/control garbage collection.
+	 */
+	public GarbageCollectorControlInterceptor(Heap monitor, Executor executor) {
+		this.monitor = monitor;
+		this.executor = executor;
+		this.sampler = new Sampler(SAMPLE_HISTORY_SIZE);
+		this.doingGC = new AtomicBoolean(false);
+		this.incoming = new AtomicLong();
+		this.finished = new AtomicLong();
+		this.shedRequests = new AtomicLong();;
+	}
+
+	/**
+	 * Creates a new instance of {@code GarbageCollectorControlInterceptor} using
+	 * defaults.
+	 *
+	 * @see Heap
+	 * @see System#gc()
+	 * @see Executors#newSingleThreadExecutor()
+	 * @see UnavailabilityDuration
+	 * @see Clock#systemDefaultZone()
+	 */
+	public GarbageCollectorControlInterceptor() {
+		this(new Heap(), Executors.newSingleThreadExecutor());
+	}
+
+	private ShedResponse shed() {
+		shedRequests.incrementAndGet();
+		return new ShedResponse(true);
+	}
+
+	public ShedResponse before() {
+		// The service is unavailable.
+		if (doingGC.get()) {
+			return shed();
+		}
+		
+		if ((incoming.get() + 1) % sampler.getCurrentSampleSize() == 0) {
+			if (monitor.getHeapUsageSinceLastGC() > sheddingThreshold.get()) {
+				// Starting unavailability period.
+				if (doingGC.get()) {
+					return shed();
+				}
+				executor.execute(() -> {
+					// Loop waiting for the queue to get empty.
+					while (finished.get() < incoming.get()) {
```

Comment:
In java you don´t need to loop. You could [wait and notify](https://www.journaldev.com/1037/java-thread-wait-notify-and-notifyall-example).

Reply:
We have talked about and decided to update it after.